rust/src/lib/net_state.rs

```diff
@@ -251,16 +262,86 @@ impl NetworkState {
             ifaces.gen_state_for_apply(&current.interfaces)?;
 
         add_net_state.interfaces = add_ifaces;
-        add_net_state.prop_list = vec!["interfaces"];
-
         chg_net_state.interfaces = chg_ifaces;
-        chg_net_state.prop_list = vec!["interfaces"];
-
         del_net_state.interfaces = del_ifaces;
-        del_net_state.prop_list = vec!["interfaces"];
+
+        self.include_route_changes(
+            &mut add_net_state,
+            &mut chg_net_state,
+            current,
+        );
 
         Ok((add_net_state, chg_net_state, del_net_state))
     }
+
+    fn include_route_changes(
+        &self,
+        add_net_state: &mut Self,
+        chg_net_state: &mut Self,
+        current: &Self,
+    ) {
+        let mut changed_iface_routes =
+            self.routes.gen_changed_ifaces_and_routes(&current.routes);
+
+        // Clippy is wrong there, we cannot use iterator here. As the
+        // add_net_state will be changed on next call of
+        // `add_iface_names.contains()`.
+        #[allow(clippy::needless_collect)]
+        let add_iface_names: Vec<String> = add_net_state
+            .interfaces
+            .kernel_ifaces
+            .keys()
+            .map(String::clone)
+            .collect();
+        // Clippy is wrong there, we cannot use iterator here. As the
+        // chg_net_state will be changed on next call of
+        // `chg_iface_names.contains()`.
+        #[allow(clippy::needless_collect)]
+        let chg_iface_names: Vec<String> = chg_net_state
+            .interfaces
+            .kernel_ifaces
+            .keys()
+            .map(String::clone)
+            .collect();
+
+        for (iface_name, routes) in changed_iface_routes.drain() {
+            if add_iface_names.contains(&iface_name) {
+                add_net_state
+                    .interfaces
+                    .mark_iface_route_changed(iface_name.as_str());
+                add_net_state.routes.extent_config_routes(routes)
+            } else if chg_iface_names.contains(&iface_name) {
+                chg_net_state
+                    .interfaces
+                    .mark_iface_route_changed(iface_name.as_str());
+                chg_net_state.routes.extent_config_routes(routes)
+            } else {
+                // Interface not mentioned in desire state but impacted by
+                // wildcard absent route
+                if let Some(cur_iface) = current
+                    .interfaces
+                    .get_iface(&iface_name, InterfaceType::Unknown)
+                {
+                    chg_net_state.routes.extent_config_routes(routes);
+                    let mut new_iface = cur_iface.clone_name_type_only();
+                    new_iface.base_iface_mut().ipv4 =
+                        cur_iface.base_iface().ipv4.clone();
+                    new_iface.base_iface_mut().ipv6 =
+                        cur_iface.base_iface().ipv6.clone();
```

Comment:
Cloning the whole current interface will cause backend crate all settings of this interface which user never asked. We set IPv4 and IPv6 because route require them, this is sort of user request.